internal/deploy/callback/callback_addon.go

```diff
@@ -159,3 +159,157 @@ func NewCallbackServer(t ct.TestLike, hostnameRunningComplement string) (*Callba
 
 	return callbackServer, nil
 }
+
+// SendError returns a callback.Fn which returns the provided statusCode
+// along with a JSON error $count times, after which it lets the response
+// pass through. This is useful for testing retries.
+func SendError(count uint32, statusCode int) Fn {
+	var seen atomic.Uint32
+	return func(d Data) *Response {
+		next := seen.Add(1)
+		if next > count {
+			return nil
+		}
+		return &Response{
+			RespondStatusCode: statusCode,
+			RespondBody:       json.RawMessage(`{"error":"callback.SendError"}`),
+		}
+	}
+}
+
+// TODO: helpers for "wait for this conditional to pass then execute this code whilst blocking"
+//  - for tarpitting
+//  - for sigkilling
+
+type PassiveChannel struct {
+	recvCh  chan *Data
+	timeout time.Duration
+	closed  *atomic.Bool
+}
+
+func (c *PassiveChannel) Close() {
+	if c.closed.CompareAndSwap(false, true) {
+		close(c.recvCh)
+	}
+}
+
+// Callback returns the callback implementation used to send data to this channel.
+func (c *PassiveChannel) Callback() Fn {
+	return func(d Data) *Response {
+		if c.closed.Load() {
```

Comment:
Is there a race here where the channel might be closed immediately after this check? Not sure what the rules are in Go.

Reply:
Bleh, you're right. Will fix.